java/cuvs-java/src/main/java22/com/nvidia/cuvs/spi/JDKProvider.java

```diff
@@ -188,4 +224,93 @@ public CuVSMatrix newMatrixFromArray(byte[][] vectors) {
+   * device memory. It uses a non-native {@link MemorySegment} created directly from on-heap java arrays to avoid
+   * an intermediate allocation and copy to a native (off-heap) segment.
+   * It requires the copy function ({@code cudaMemcpyAsync}) to have the {@code Critical} linker option in order
+   * to allow the access to on-heap memory (see {@link Native#cudaMemcpyAsync}).
+   */
+  private static class HeapSegmentBuilder implements CuVSMatrix.Builder<CuVSDeviceMatrix> {
+    private final long columns;
+    private final long size;
+    private final CuVSDeviceMatrixImpl matrix;
+    private final MemorySegment stream;
+    int current;
+
+    private HeapSegmentBuilder(
+        CuVSResources resources, long size, long columns, CuVSMatrix.DataType dataType) {
+      this.columns = columns;
+      this.size = size;
+      this.matrix = CuVSDeviceMatrixRMMImpl.create(resources, size, columns, dataType);
+      this.stream = Util.getStream(resources);
+      current = 0;
+    }
+
+    private HeapSegmentBuilder(
+        CuVSResources resources,
+        long size,
+        long columns,
+        int rowStride,
+        int columnStride,
+        CuVSMatrix.DataType dataType) {
+      this.columns = columns;
+      this.size = size;
+      this.matrix =
+          CuVSDeviceMatrixRMMImpl.create(
+              resources, size, columns, rowStride, columnStride, dataType);
+      this.stream = Util.getStream(resources);
+      current = 0;
+    }
+
+    @Override
+    public void addVector(float[] vector) {
+      if (vector.length != columns) {
+        throw new IllegalArgumentException(
+            String.format(
+                Locale.ROOT, "Expected a vector of size [%d], got [%d]", columns, vector.length));
+      }
+      internalAddVector(MemorySegment.ofArray(vector));
+    }
+
+    @Override
+    public void addVector(byte[] vector) {
+      if (vector.length != columns) {
+        throw new IllegalArgumentException(
+            String.format(
+                Locale.ROOT, "Expected a vector of size [%d], got [%d]", columns, vector.length));
+      }
+      internalAddVector(MemorySegment.ofArray(vector));
+    }
+
+    @Override
+    public void addVector(int[] vector) {
+      if (vector.length != columns) {
+        throw new IllegalArgumentException(
+            String.format(
+                Locale.ROOT, "Expected a vector of size [%d], got [%d]", columns, vector.length));
+      }
+      internalAddVector(MemorySegment.ofArray(vector));
+    }
+
+    private void internalAddVector(MemorySegment vector) {
+      if (current >= size) {
+        throw new ArrayIndexOutOfBoundsException();
+      }
+
+      long rowBytes = columns * matrix.valueLayout().byteSize();
+
+      var dstOffset = ((current++) * rowBytes);
+      var dst = matrix.memorySegment().asSlice(dstOffset);
+      checkCudaError(
+          Native.cudaMemcpyAsync(dst, vector, rowBytes, cudaMemcpyHostToDevice(), stream),
+          "cudaMemcpyAsync");
```

Comment:
This might warrant discussion.  I might not have a good handle on the call stack or usage pattern for this function.

As I understand it, the `vector` has of the order of 1000s of elements?  Is this to be called occasionally, or is this the regular control flow?

You're likely aware:  Calling `cudaMemcpy`/`cudaMemcpyAsync` in a loop for kilobytes at a time is going to have much lower throughput than doing a block copy.